BUG: contour plot levels

fmaussion opened the pull request:
This should fix https://github.com/pydata/xarray/issues/866

fmaussion commented on xarray/plot/plot.py:
```diff
@@ -387,6 +387,18 @@ def newplotfunc(darray, x=None, y=None, ax=None, row=None, col=None,
+                if len(levels) < 2:
+                    add_colorbar = False
+            except TypeError:
+                if levels < 2:
```
This is a huge corner case, but well...

shoyer commented on xarray/plot/plot.py:
```diff
@@ -384,8 +384,16 @@ def newplotfunc(darray, x=None, y=None, ax=None, row=None, col=None,
+            if levels is None:
+                levels = 7  # this is the matplotlib default
+            # A colorbar with one level is not possible with mpl
+            try:
```
Can we do a type check instead? e.g.,
```
if (isinstance(levels, int) and levels < 2) or len(levels) < 2:
    add_colorbar = False
```

shoyer replied:
We could also change the default to `add_colorbar=None`, and only apply this logic if `add_colorbar` wasn't set directly.

shoyer commented:
I wonder -- would it maybe make more sense to simply *not* to add a colorbar by default? It's pretty easy to toggle if desired.

fmaussion commented:
@shoyer yes, this was my original idea too: `contour` plots usually don't need a colorbar. I changed the default to false for `contour` plots only, and removed the convoluted logic for when the user gives only one level: this is a matplotlib problem after all, and if the user sets `levels=1` and `add_colorbar=True` it really isn't xarray's problem anymore.

I think it's better now, let me know what you think.

shoyer commented on xarray/plot/utils.py:
```diff
@@ -41,7 +41,7 @@ def _build_discrete_cmap(cmap, levels, extend, filled):
     if not filled:
         # non-filled contour plots
-        extend = 'neither'
+        extend = 'max'
```
I think I missed the rationale for this one -- why should this change?

fmaussion replied:
This is the actual bug fix for https://github.com/pydata/xarray/issues/866 which motivated this PR ;-).

xarray wasn't plotting contours correctly before this change, see the following code:

```python
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr

x, y = np.meshgrid(np.arange(12), np.arange(12))
z = xr.DataArray(np.sqrt(x**2 + y**2))
z.plot.contour(levels=[2, 4, 6, 8], add_colorbar=False, colors='k')
plt.show()
```
![figure_1](https://cloud.githubusercontent.com/assets/10050469/16994818/5779881c-4ea9-11e6-802c-5a0d671517e2.png)

With this fix everything is fine now:

![figure_1-1](https://cloud.githubusercontent.com/assets/10050469/16994804/45776abc-4ea9-11e6-8b19-d1a05a7d1103.png)

Intuitively, it also makes sense: the last contour does mark all data beyond the last level, hence `extend = 'max'`

shoyer replied:
Got it, thanks!
